deepinv/physics/phase_retrieval.py

```diff
@@ -153,9 +278,231 @@ def __init__(
             device=device,
             rng=self.rng,
         )
-        super().__init__(B)
-        self.name = f"RPR_m{self.m}"
+        super().__init__(B, **kwargs)
+        self.name = "Random Phase Retrieval"
+
+    def get_A_squared_mean(self):
+        return self.B._A.var() + self.B._A.mean() ** 2
+
+
+class StructuredRandomLinearOperator(LinearPhysics):
+    r"""
+    Linear operator for structured random phase retrieval.
+
+    :param tuple input_shape: shape (C, H, W) of inputs.
+    :param tuple output_shape: shape (C, H, W) of outputs.
+    :param str mode: sampling mode.
+    :param float n_layers: number of layers.
+    :param function transform_func: structured transform function.
+    :param function transform_func_inv: structured inverse transform function.
+    :param list diagonals: list of diagonal matrices.
+    """
+
+    def __init__(
+        self,
+        input_shape,
+        output_shape,
+        mode,
+        n_layers,
+        transform_func,
+        transform_func_inv,
+        diagonals,
+        **kwargs,
+    ):
+
+        def A(x):
+            assert (
+                x.shape[1:] == input_shape
+            ), f"x doesn't have the correct shape {x.shape[1:]} != {input_shape}"
+
+            if mode == "oversampling":
+                x = padding(x, input_shape, output_shape)
+
+            if n_layers - math.floor(n_layers) == 0.5:
+                x = transform_func(x)
+            for i in range(math.floor(n_layers)):
+                diagonal = diagonals[i]
+                x = diagonal * x
+                x = transform_func(x)
+
+            if mode == "undersampling":
+                x = trimming(x, input_shape, output_shape)
+
+            return x
+
+        def A_adjoint(y):
+            assert (
+                y.shape[1:] == output_shape
+            ), f"y doesn't have the correct shape {y.shape[1:]} != {self.output_shape}"
+
+            if mode == "undersampling":
+                y = padding(y, input_shape, output_shape)
+
+            for i in range(math.floor(n_layers)):
+                diagonal = diagonals[-i - 1]
+                y = transform_func_inv(y)
+                y = torch.conj(diagonal) * y
+            if n_layers - math.floor(n_layers) == 0.5:
+                y = transform_func_inv(y)
+
+            if mode == "oversampling":
+                y = trimming(y, input_shape, output_shape)
+
+            return y
+
+        super().__init__(A=A, A_adjoint=A_adjoint, **kwargs)
+
+
+class StructuredRandomPhaseRetrieval(PhaseRetrieval):
+    r"""
+    Structured random phase retrieval model corresponding to the operator
+
+    .. math::
+
+        A(x) = |\prod_{i=1}^N (F D_i) x|^2,
+
+    where :math:`F` is the Discrete Fourier Transform (DFT) matrix, and :math:`D_i` are diagonal matrices with elements of unit norm and random phases, and :math:`N` refers to the number of layers. It is also possible to replace :math:`x` with :math:`Fx` as an additional 0.5 layer.
+
```

Comment:
It is not clear to me how is the subsampling/oversampling achieved from the description of the operator. Can you explain this in more detail?